doc/shared/sunadjoint/SUNAdjointStepper.rst

```diff
@@ -0,0 +1,247 @@
+..
+   ----------------------------------------------------------------
+   SUNDIALS Copyright Start
+   Copyright (c) 2002-2024, Lawrence Livermore National Security
+   and Southern Methodist University.
+   All rights reserved.
+
+   See the top-level LICENSE and NOTICE files for details.
+
+   SPDX-License-Identifier: BSD-3-Clause
+   SUNDIALS Copyright End
+   ----------------------------------------------------------------
+
+.. _SUNAdjointStepper:
+
+The SUNAdjointStepper API and Module
+====================================
+
+The :c:type:`SUNAdjointStepper` API and module provides a package-agnostic
+interface to SUNDIALS ASA capabilities. It currently only supports the discrete
+ASA capabilities in the ARKODE package, but in the future this support may be expanded.
+
+A :c:type:`SUNAdjointStepper` is a pointer to the
+:c:struct:`SUNAdjointStepper_` structure:
+
+.. c:type:: struct SUNAdjointStepper_ *SUNAdjointStepper
+
+.. c:struct:: SUNAdjointStepper_
+
+   .. c:member:: SUNStepper adj_sunstepper
+
+      The :c:type:`SUNStepper` object used for backwards time stepping of the adjoint ODE system.
+
+   .. c:member:: SUNStepper fwd_sunstepper
+
+      The :c:type:`SUNStepper` object used for forward time stepping of the original ODE system if any recomputation of missing
+      state data is required during the backwards integration.
+
+   .. c:member:: sunrealtype tf
```

Comment:
Its stored because the `SUNAdjointStepper_Evolve` function does not take `tf` as an argument like `SUNStepper_Evolve`. I could change `SUNAdjointStepper_Evolve` to have a `tf` argument and get rid of the need to store it.